### ISSUE TITLE
Lower level API?

### ISSUE THREAD

**Issue by spion** [rustformers/llm#267]
I was wondering if any of the crates exposes a simple, low-level API

Something along the lines of

`(sessionState, tokenIds) -> (nextSessionState, next_logits)`

This separates the network itself from the other fiddly bits, allowing you to implement any strategy for handling the logits (such as increasing/decreasing probabilities based on external criteria before sampling with top_p / top_k, restricting to a subset with logits using context-free grammars, or any number of other strategies i.e. https://github.com/rustformers/llm/issues/235 stuff). 

It also removes the need to pass a rng, have a lot of parameters tied to a session, lets one use a different tokenizer (e.g. the huggingface one) and removes the need to set up progress tracking callbacks.

**Comment by tanmaysachan** [comment]
I'll try to get an initial implementation done for this.

**Comment by philpax** [comment]
Hi there! I've just merged #280 which should address this by letting you define your own sampling strategy, which you can combine with `infer_next_token`. Can you let me know if that solves your problem?

The reason for not supporting arbitrary inference is that the previous tokens are required for continued sampling, so the session needs to know which token was chosen after inference.

**Comment by spion** [comment]
The Sampler trait looks perfect at first glance, I'm going to give it a try in my project.
